src/modules/python/pbs/v1/_pmi_cray.py

```diff
@@ -0,0 +1,424 @@
+        except:
+            pass
+    return nidset
+
+
+def nodenids(hosts):
+    # return the set of nids from the host list
+    nidset = set()
+    craynid = "PBScraynid"
+    for vnames in hosts:
+        vnode = _svr_vnode(vnames)
+        try:
+            nidset.add(int(vnode.resources_available[craynid]))
+        except:
+            pass
+    return nidset
+
+
+def nidlist(job, nidset=None):
+    # Return a string to be used with capmc --nids option
+    # and the number of nids
+
+    if nidset is None:
+        nidset = jobnids(job)
+    nids = []
+    first = ""
+    last = ""
+    prev = 0
+    for nid in sorted(nidset):
+        val = nid
+        if len(first) == 0:		# start point
+            first = str(nid)
+        elif prev + 1 != val:
+            if prev == int(first):
+                nids.append(first)
+            else:
+                nids.append(first + "-" + last)
+            first = str(nid)
+        prev = val
+        last = str(nid)
+    if len(first) > 0:
+        if prev == int(first):
+            nids.append(first)
+        else:
+            nids.append(first + "-" + last)
+    return ",".join(nids), len(nidset)
+
+
+def spool_file(name):
+    # Form a path to the PBS spool directory with name as the last element.
+    home = _pbs_conf("PBS_HOME")
+    if home is None:
+        raise BackendError("PBS_HOME not found")
+    return os.path.join(home, "spool", name)
+
+
+def energy_file(job):
+    return spool_file("%s.energy" % job.id)
+
+
+def rur_file(job):
+    return spool_file("%s.rur" % job.id)
+
+
+def node_energy(id, nids, cnt):
+    # Return the result of running capmc get_node_energy_counter.
+    # The magic number of 15 seconds in the past is used because that
+    # is the most current value that can be expected from capmc.
+    if cnt == 0:
+        return None
+    cmd = "get_node_energy_counter --nids %s" % nids
+    ret = launch(id, cmd)
+    cntkey = "nid_count"
+    gotcnt = "<notset>"
+    if (ret is not None) and (cntkey in ret):
+        gotcnt = ret[cntkey]
+        if gotcnt == cnt:
+            return ret
+
+    pbs.logjobmsg(id, "node count %s, should be %d" % (str(gotcnt), cnt))
+    ret = launch(id, cmd)
+    gotcnt = "<notset>"
+    if (ret is not None) and (cntkey in ret):
+        gotcnt = ret[cntkey]
+        if gotcnt == cnt:
+            return ret
+
+    pbs.logjobmsg(id, "second query failed, node count %s, should be %d" %
+                      (str(gotcnt), cnt))
+    return None
+
+
+def job_energy(job, nids, cnt):
+    # Return energy counter from capmc.  Return None if no energy
+    # value is available.
+    energy = None
+    ret = node_energy(job.id, nids, cnt)
+    if ret is not None and "nodes" in ret:
+        energy = 0
+        for node in ret["nodes"]:
+            energy += node["energy_ctr"]
+        pbs.logjobmsg(job.id, "energy usage %dJ" % energy)
+    return energy
+
+
+class Pmi:
+
+    def __init__(self, pyhome=None):
```

Comment:
If pyhome is not used in this function, remove it as argument.

Reply:
It is kept to match the generic api which is also used for SGI.